### ISSUE TITLE
cross_batch_memory The implementation does not appear to be a queue data structure

### ISSUE THREAD

**Issue by qianjinhao** [KevinMusgrave/pytorch-metric-learning#54]
`        if end_idx > self.queue_idx:
            self.embedding_memory[self.queue_idx:end_idx] = embeddings.detach()
            self.label_memory[self.queue_idx:end_idx] = labels.detach()            
        else:
            se = self.memory_size-self.queue_idx
            self.embedding_memory[self.queue_idx:] = embeddings[:se].detach()
            self.embedding_memory[:end_idx] = embeddings[se:].detach()
            self.label_memory[self.queue_idx:] = labels[:se].detach()
            self.label_memory[:end_idx] = labels[se:].detach()`

When the memory is full, the queue head is always replaced, but the queue head is replaced with the most recent batch, so the last half of the queue remains unchanged, without the first-in-first-out feature of the queue.

**Comment by KevinMusgrave** [comment]
Hmm, but ```self.queue_idx``` is moved forward every iteration:
```python
self.queue_idx = (self.queue_idx + batch_size) % self.memory_size
```

**Comment by qianjinhao** [comment]
> Hmm, but `self.queue_idx` is moved forward every iteration:
> 
> ```python
> self.queue_idx = (self.queue_idx + batch_size) % self.memory_size
> ```

I understand now, thanks for your reply
